### PR TITLE
ELECTRON-600 (Implement Native Encryption for AESGCMEncrypt, AESGCMDecrypt & RSADecrypt)

### PR DESCRIPTION
**Pull request by KiranNiranjan**
## Description
Implement Native Encryption for AESGCMEncrypt, AESGCMDecrypt & RSADecrypt [ELECTRON-600](https://perzoinc.atlassian.net/browse/ELECTRON-600)

~~Also, as a side note, we aren't supporting RSA related algorithms yet due to crypto-lib issues, we are working with the security & mobile teams to find out how to resolve this and how the RSA related algorithms are being used. We'll raise a separate PR for that once those issues are resolved.~~

## Related PRs
List related PRs against other branches/repositories:

branch | PR
------ | ------
SFE-Packages | [link]()
SFE-Client-App | [link]()

## QA Checklist
- [X] Unit-Tests
[ELECTRON-600 — Unit Tests.pdf](https://github.com/symphonyoss/SymphonyElectron/files/2262019/ELECRON-600.Unit.Tests.pdf)

- [X] Automation-Tests
[ELECTRON-600 — Spectron.pdf](https://github.com/symphonyoss/SymphonyElectron/files/2262017/ELECTRON-600.Spectron.pdf)

### CODE REVIEW

**Comment by VishwasShashidhar** [comment]
@KiranNiranjan can you look at the conflicts please?

File: demo/index.html [hunk 1]
```diff
@@ -278,5 +286,32 @@ <h1>Symphony Electron API Demo</h1>
...
+        encrypt.addEventListener('click', function () {
+            encryptedResult.innerText = ssf.CryptoLib.AESGCMEncrypt(GCM_VECTOR_16.IV, GCM_VECTOR_16.A, GCM_VECTOR_16.K, GCM_VECTOR_16.P);
+        });
+
+        var decrypt = document.getElementById('decrypt');
+        var decryptedResult = document.getElementById('decrypted-result');
+        decrypt.addEventListener('click', function () {
+            decryptedResult.innerText = ssf.CryptoLib.AESGCMDecrypt(GCM_VECTOR_16.IV, GCM_VECTOR_16.A, GCM_VECTOR_16.K, realCt);
+        });
+
```
Comment (VikasShashidhar) [inline demo/index.html]: Let's not have anything on the demo page related to security. Please remove it.

File: js/cryptoLib.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        ref.types.uint32,
+        ref.refType(ref.types.uchar),
+    ]],
+
+    getVersion: [ref.types.CString, []],
+});
+
+const AESGCMEncrypt = function(Base64IV, Base64AAD, Base64Key, Base64In) {
+
+    const TagLen = TAG_LENGTH;
```
Comment (VikasShashidhar) [inline js/cryptoLib.js]: Why the need for a redundant variable here?

File: js/cryptoLib.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    const IV = Buffer.from(Base64IV, 'base64');
+    const AAD = Buffer.from(Base64AAD, 'base64');
+    const Key = Buffer.from(Base64Key, 'base64');
+    const In = Buffer.from(base64In, 'base64');
+
+    const OutPtr = Buffer.alloc(In.length);
+    const Tag = Buffer.alloc(TAG_LENGTH);
+
+    const ret = library.AESEncryptGCM(In, In.length, AAD, AAD.length, Key, IV, IV.length, OutPtr, Tag, TagLen);
```
Comment (VikasShashidhar) [inline js/cryptoLib.js]: Please change variable name to something more appropriate.

File: js/cryptoLib.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    const Tag = Buffer.alloc(TAG_LENGTH);
+
+    const ret = library.AESDecryptGCM(In, CipherTextLen, AAD, AAD.length, Tag, TagLen, Key, IV, IV.length, OutPtr);
+
+    if (ret < 0) {
+        log.send(logLevels.ERROR, `AESDecryptGCM, Failed to decrypt with exit code ${ret}`);
+    }
+
+    log.send(logLevels.INFO, `Output from AESDecryptGCM ${ret}`);
+
```
Comment (VikasShashidhar) [inline js/cryptoLib.js]: Unneeded empty line.

File: js/cryptoLib.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    const AAD = Buffer.from(Base64AAD, 'base64');
+    const Key = Buffer.from(Base64Key, 'base64');
+    const In = Buffer.from(base64In, 'base64');
+
+    const OutPtr = Buffer.alloc(In.length);
+    const Tag = Buffer.alloc(TAG_LENGTH);
+
+    const resultCode = library.AESEncryptGCM(In, In.length, AAD, AAD.length, Key, IV, IV.length, OutPtr, Tag, TAG_LENGTH);
+
+    if (resultCode < 0) {
```
Comment (VishwasShashidhar) [inline js/cryptoLib.js]: if `resultCode < 0`, shouldn't we also exit this function?

File: js/cryptoLib.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    const Key = Buffer.from(Base64Key, 'base64');
+    const In = Buffer.from(base64In, 'base64');
+
+    const OutPtr = Buffer.alloc(In.length - TAG_LENGTH);
+    const CipherTextLen = In.length - TAG_LENGTH;
+    const Tag = Buffer.alloc(TAG_LENGTH);
+
+    const resultCode = library.AESDecryptGCM(In, CipherTextLen, AAD, AAD.length, Tag, TAG_LENGTH, Key, IV, IV.length, OutPtr);
+
+    if (resultCode < 0) {
```
Comment (VishwasShashidhar) [inline js/cryptoLib.js]: if `resultCode < 0`, shouldn't we also exit this function?

File: js/cryptoLib.js [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+const RSADecrypt = function (pemKey, input) {
+    return RSAEncryptDecrypt("RSADecrypt", pemKey, input);
+};
+
+const RSAEncryptDecrypt = function (action, pemKey, inputStr) {
+
+    let rsaKey = getRSAKeyFromPEM(pemKey);
+
+    if (!rsaKey) {
```
Comment (VishwasShashidhar) [inline js/cryptoLib.js]: if `!rsaKey, shouldn't we also exit this function?

File: js/cryptoLib.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        log.send(logLevels.ERROR, `Failed to parse formatted RSA PEM key -> ${pemKey}`);
+    }
+
+    let input = Buffer.from(inputStr, 'base64');
+    let outLen = library.getRSAKeySize(rsaKey);
+
+    let outPtr = Buffer.alloc(32);
```
Comment (VishwasShashidhar) [inline js/cryptoLib.js]: please move `32` to a variable

**Pull request merged**